dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1328,6 +1331,108 @@ private List<Event> inlineMemCpy(FunctionCall call) {
         return replacement;
     }
 
+    // https://en.cppreference.com/w/c/string/byte/memcpy
+    private List<Event> inlineMemCpyS(FunctionCall call) {
+        // Cast guaranteed to success by the return type of memcpy_s
+        final Register resultRegister = ((ValueFunctionCall)call).getResultRegister();
+        final Function caller = call.getFunction();
+        final Expression dest = call.getArguments().get(0);
+        final Expression destszExpr = call.getArguments().get(1);
+        final Expression src = call.getArguments().get(2);
+        final Expression countExpr = call.getArguments().get(3);
+
+        if (!(countExpr instanceof IntLiteral countValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic count argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int count = countValue.getValueAsInt();
+        if (!(destszExpr instanceof IntLiteral destszValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic destsz argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int destsz = destszValue.getValueAsInt();
+
+        // Runtime checks
+        final Expression nullExpr = expressions.makeZero(types.getArchType());
+        final Expression destIsNull = expressions.makeEQ(dest, nullExpr);
+        final Expression srcIsNull = expressions.makeEQ(src, nullExpr);
+        // We assume RSIZE_MAX = 2^64-1
+        final Expression rsize_max = expressions.makeValue(BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE), types.getArchType());
+        final Expression invalidDestsz = expressions.makeGT(
+                expressions.makeCast(destszExpr, types.getArchType()), rsize_max, false);
+        final Expression countGtMax = expressions.makeGT(
+                expressions.makeCast(countExpr, types.getArchType()), rsize_max, false);
+        final Expression countGtdestszExpr = expressions.makeGT(countExpr, destszExpr, false);
+        final Expression invalidCount = expressions.makeOr(countGtMax, countGtdestszExpr);
+        // src and dest are pointers (int64), count is usually interpreted as int32, thus the cast
```

Comment:
This and the other casts in the pass really bother me. Without them, I was getting different BVs sizes, but I think the compiler and the type system should take care of giving me the correct types. So there is something that still seems fishy here. I suspect my system might not be properly recognizing `memcpy_s`

Reply:
You can see this problem in your `.ll` file:
```
declare dso_local i32 @memcpy_s(...) #3 // varargs declaration because the function is undeclared in the C-code
//... The call is thus a varags call with ugly function type casts:
  %73 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %71, i64 12, i32* %72, i64 12), !dbg !86
```
This function is declared in `string.h` according to cppreference and you are not including that one.

Reply:
Even if I add `string.h`, function `memcpy_s` ends up with an empty body in the llvm file. There is this

> As with all bounds-checked functions, memcpy_s only guaranteed to be available if __STDC_LIB_EXT1__ is defined by the implementation and if the user defines __STDC_WANT_LIB_EXT1__ to the integer constant 1 before including [<string.h>](https://en.cppreference.com/w/c/string/byte).

and by trying the code from [here](https://stackoverflow.com/questions/47867130/stdc-lib-ext1-availability-in-gcc-and-clang), I get `#error "__STDC_LIB_EXT1__ not defined"`, so we might need a solution that works even if the implementation of `memcpy_s` is not available.

However, if I remove the casts from the code, I am now not able to reproduce the parsing problems I was getting before. What I don't understand is that even if the llvm code show different types in the call arguements
```
  %8 = call i32 (i8*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i8*, i64, i32*, i64, ...)*)(i8* null, i64 12, i32* %7, i64 12), !dbg !27
  %15 = call i32 (i32*, i64, i8*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i8*, i64, ...)*)(i32* %14, i64 12, i8* null, i64 12), !dbg !34
  %44 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %42, i64 8, i32* %43, i64 16), !dbg !60
  %73 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %71, i64 12, i32* %72, i64 12), !dbg !86
  %102 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %100, i64 12, i32* %101, i64 12), !dbg !112
```
we parse all of them as bv64
```
11:      bv32 r8 <- call memcpy_s(bv64(0), bv64(12), bv64 r7, bv64(12))
30:      bv32 r15 <- call memcpy_s(bv64 r14, bv64(12), bv64(0), bv64(12))
107:     bv32 r44 <- call memcpy_s(bv64 r42, bv64(8), bv64 r43, bv64(16))
184:     bv32 r73 <- call memcpy_s(bv64 r71, bv64(12), bv64 r72, bv64(12))
261:     bv32 r102 <- call memcpy_s(bv64 r100, bv64(12), bv64 r101, bv64(12))
```

Reply:
> The count/size parameters are more dangerous. Since your code uses `sizeof(...)` everywhere, you will get a consistent type `size_t` which equals 64 bit on most systems. If you were to put constants that are interpreted as 32 bit ints, then you would see the problem again. 

The type of those is `rsize_t` which AFAIK is defined as `typedef size_t rsize_t;`. Thus, the compiled should take care of the casting even if constant are used. However, for 32 bits archs, we probably get into problems.

> Always upcasting those parameters to 64 bit is probably fine though.

Since we overall assume 64 bits arch, I guess we should use the casts. I will put those two back.

> 
> EDIT: Btw. even in your newest .ll file, the function is not properly recognized.

If you mean that we have `declare dso_local i32 @memcpy_s(...) #3` with no body, I am aware of this. This is what I meant above with `"__STDC_LIB_EXT1__ not defined"` and apparently is not mandatory to have them implemented

> These functions are defined in ISO 9899:2011 Annex K and they are optional to implement. Before using them you must check if __STDC_LIB_EXT1__ is defined.

Reply:
I'm not talking specifically about the body, but the whole signature. Since you had a commit to include `string.h` and to regenerate the llvm file, I thought you found a way to get this function recognized.
It might make sense you just declare the function in the benchmark to get the right signature (not strictly necessary though).